Preventing int64 out of range panics on the Resampler

In Resamplers, a slice can be indexed with over `9223372036854775807` unchecked, which causes a panic. An error returned is obviously better than `panic()`

 Original issue: https://github.com/faiface/beep/issues/97

I've tested @skuzzymiglet's application (which has been archived but is awesome still) and when I pass in a tempo of 100000, internally it calculates a ratio to be `+Inf` and the `Resampler` doesn't know what to do with that.

MVP:
```golang
package main

import (
	"github.com/gopxl/beep"
	"github.com/gopxl/beep/generators"
	"github.com/gopxl/beep/speaker"
	"math"
)

func main() {
	s, err := generators.SineTone(44100, 261.63)
	if err != nil {
		panic(err)
	}

	rs := beep.ResampleRatio(10, math.Inf(1), s)

	speaker.Init(44100, 512)
	done := make(chan bool)
	speaker.Play(beep.Seq(beep.Take(44100*3, rs), beep.Callback(func() {
		done <- true
	})))
	<-done
}
```

```
panic: runtime error: index out of range [-9223372036854775296]

goroutine 19 [running]:
github.com/gopxl/beep.(*Resampler).Stream(0xc0000b4000, {0xc000202000?, 0x200?, 0x200?})
	/.../beep/resample.go:96 +0x3db
github.com/gopxl/beep.(*take).Stream(0xc000012018, {0xc000202000?, 0x67?, 0x400000?})
	/.../beep/compositors.go:26 +0x42
main.main.Seq.func3({0xc000202000?, 0x496e40?, 0x426201?})
	/.../beep/compositors.go:86 +0xb3
github.com/gopxl/beep.StreamerFunc.Stream(0x426280?, {0xc000202000?, 0x3c0?, 0x400?})
	/.../beep/interface.go:100 +0x24
github.com/gopxl/beep.(*Mixer).Stream(0x545ca0, {0xc000180000, 0x5622, 0x5622})
	/.../beep/mixer.go:42 +0x178
github.com/gopxl/beep/speaker.(*sampleReader).stream(0x56d1e8?, {0xc000180000?, 0x45d245?, 0x400000?})
	/.../beep/speaker/speaker.go:155 +0xc2
github.com/gopxl/beep/speaker.(*sampleReader).Read(0xc0000a40c0, {0xc000116000, 0x15888, 0xc000104df0?})
	/.../beep/speaker/speaker.go:119 +0xc5
github.com/ebitengine/oto/v3/internal/mux.(*playerImpl).read(0x7fa31e8d9f18?, {0xc000116000?, 0xc000058800?, 0xc000200000?})
	/home/markkremer/go/pkg/mod/github.com/ebitengine/oto/v3@v3.1.0/internal/mux/mux.go:268 +0xc8
github.com/ebitengine/oto/v3/internal/mux.(*playerImpl).readSourceToBuffer(0xc00010e000)
	/home/markkremer/go/pkg/mod/github.com/ebitengine/oto/v3@v3.1.0/internal/mux/mux.go:529 +0x145
github.com/ebitengine/oto/v3/internal/mux.(*Mux).loop(0xc0000b6080)
	/home/markkremer/go/pkg/mod/github.com/ebitengine/oto/v3@v3.1.0/internal/mux/mux.go:108 +0x205
created by github.com/ebitengine/oto/v3/internal/mux.New in goroutine 1
	/home/markkremer/go/pkg/mod/github.com/ebitengine/oto/v3@v3.1.0/internal/mux/mux.go:69 +0xf8
exit status 2
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 - Upgrade Go version to 1.21 ([#2](https://github.com/gopxl/beep/pull/2))
 - Upgrade Oto version to 3.1 ([#3](https://github.com/gopxl/beep/pull/3))
+- Panic when `Resampler` is given a ratio of `Inf` or `NaN`. ([#120](https://github.com/gopxl/beep/pull/120))
 
 ## [v1.0.0] 2023-10-07
 - Forked [faiface/beep](https://github.com/faiface/beep) to [gopxl/beep](https://github.com/gopxl/beep).

--- a/resample.go
+++ b/resample.go
@@ -1,6 +1,9 @@
 package beep
 
-import "fmt"
+import (
+	"fmt"
+	"math"
+)
 
 // Resample takes a Streamer which is assumed to stream at the old sample rate and returns a
 // Streamer, which streams the data from the original Streamer resampled to the new sample rate.
@@ -43,6 +46,9 @@ func Resample(quality int, old, new SampleRate, s Streamer) *Resampler {
 func ResampleRatio(quality int, ratio float64, s Streamer) *Resampler {
 	if quality < 1 || 64 < quality {
 		panic(fmt.Errorf("resample: invalid quality: %d", quality))
+	}
+	if math.IsInf(ratio, 0) || math.IsNaN(ratio) {
+		panic(fmt.Errorf("resample: invalid ratio: %d", ratio))
 	}
 	return &Resampler{
 		s:     s,
@@ -148,6 +154,9 @@ func (r *Resampler) Ratio() float64 {
 
 // SetRatio sets the resampling ratio. This does not cause any glitches in the stream.
 func (r *Resampler) SetRatio(ratio float64) {
+	if math.IsInf(ratio, 0) || math.IsNaN(ratio) {
+		panic(fmt.Errorf("resample: invalid ratio: %d", ratio))
+	}
 	r.pos = int(float64(r.pos) * r.ratio / ratio)
 	r.ratio = ratio
 }

